addons/l10n_in_ewaybill_stock/tools/ewaybill_api.py

```diff
@@ -0,0 +1,207 @@
+            if response.get('error'):
+                raise EWayBillError(response)
+        except AccessError as e:
+            _logger.warning("Connection error: %s", e.args[0])
+            return {
+                "error": [{
+                    "code": "access_error",
+                    "message": _("Unable to connect to the E-WayBill service."
+                                 "The web service may be temporary down. Please try again in a moment.")
+                }]
+            }
+        return response
+
+    def _ewaybill_check_authentication(self):
+        sudo_company = self.company.sudo()
+        if sudo_company.l10n_in_edi_ewaybill_username and sudo_company._l10n_in_edi_ewaybill_token_is_valid():
+            return True
+        elif sudo_company.l10n_in_edi_ewaybill_username and sudo_company.l10n_in_edi_ewaybill_password:
+            try:
+                self._ewaybill_authenticate()
+                return True
+            except EWayBillError:
+                return False
+        return False
+
+    def _ewaybill_authenticate(self):
+        params = {"password": self.company.sudo().l10n_in_edi_ewaybill_password}
+        response = self._ewaybill_connect_to_server(url_path="/iap/l10n_in_edi_ewaybill/1/authenticate", params=params)
+        if response and response.get("status_cd") == "1":
+            self.company.sudo().l10n_in_edi_ewaybill_auth_validity = fields.Datetime.now() + timedelta(
+                hours=6, minutes=00, seconds=00)
+        return response
+
+    def _ewaybill_make_transaction(self, operation_type, json_payload):
+        """
+        :params operation_type: operation_type must be strictly `generate` or `cancel`
+        :params json_payload: to be sent as params
+        This method handles the common errors in generating and canceling the ewaybill
+        """
+        if not self._ewaybill_check_authentication():
+            return self._ewaybill_no_config_response()
+        params = {"json_payload": json_payload}
+        url_path = f"/iap/l10n_in_edi_ewaybill/1/{operation_type}"
+        try:
+            response = self._ewaybill_connect_to_server(
+                url_path=url_path,
+                params=params
+            )
+            return response
+        except EWayBillError as e:
+            if "no-credit" in e.error_codes:
+                e.error_json.get('odoo_warning').append({
+                    'message': self.env['account.edi.format']._l10n_in_edi_get_iap_buy_credits_message(self.company)
+                })
+                return e
+
+            if '238' in e.error_codes:
+                # Invalid token eror then create new token and send generate request again.
+                # This happens when authenticate called from another odoo instance with same credentials
+                # (like. Demo/Test)
+                with contextlib.suppress(EWayBillError):
+                    self._ewaybill_authenticate()
```

Comment:
As per @jva-odoo the authenticate request, doesn't gives us a proper error if not wrong and since it is a force authentication and re-calling the main request better to pass the error which we get from the main request.

Reply:
If the user really has a problem with the authentication it will just ask to authenticate in the next call?